internal/iam/repository_role.go

```diff
@@ -184,16 +187,141 @@ func (r *Repository) DeleteRole(ctx context.Context, withPublicId string, _ ...O
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing scope id")
 	}
-	var roles []*Role
-	err := r.list(ctx, &roles, "scope_id in (?)", []any{withScopeIds}, opt...)
+	opts := getOpts(opt...)
+
+	limit := r.defaultLimit
+	if opts.withLimit != 0 {
+		// non-zero signals an override of the default limit for the repo.
+		limit = opts.withLimit
+	}
+
+	var args []any
+	whereClause := "scope_id in @scope_ids"
+	args = append(args, sql.Named("scope_ids", withScopeIds))
+
+	query := fmt.Sprintf(listRolesTemplate, whereClause, limit)
+	if opts.withStartPageAfterItem != nil {
+		query = fmt.Sprintf(listRolesPageTemplate, whereClause, limit)
+		args = append(args,
+			sql.Named("last_item_create_time", opts.withStartPageAfterItem.GetCreateTime()),
+			sql.Named("last_item_id", opts.withStartPageAfterItem.GetPublicId()),
+		)
+	}
+	return r.queryRoles(ctx, query, args, limit)
+}
+
+// listRolesRefresh lists roles in the given scopes and supports the
+// WithLimit and WithStartPageAfterItem options.
+func (r *Repository) listRolesRefresh(ctx context.Context, updatedAfter time.Time, withScopeIds []string, opt ...Option) ([]*Role, time.Time, error) {
+	const op = "iam.(Repository).listRolesRefresh"
+
+	switch {
+	case updatedAfter.IsZero():
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing updated after time")
+
+	case len(withScopeIds) == 0:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing scope id")
+	}
+
+	opts := getOpts(opt...)
+
+	limit := r.defaultLimit
+	if opts.withLimit != 0 {
+		// non-zero signals an override of the default limit for the repo.
+		limit = opts.withLimit
+	}
+
+	var args []any
+	whereClause := "scope_id in @scope_ids"
+	args = append(args, sql.Named("scope_ids", withScopeIds))
+
+	query := fmt.Sprintf(refreshRolesTemplate, whereClause, limit)
+	args = append(args,
+		sql.Named("updated_after_time", timestamp.New(updatedAfter)),
+	)
+	if opts.withStartPageAfterItem != nil {
+		query = fmt.Sprintf(refreshRolesPageTemplate, whereClause, limit)
+		args = append(args,
+			sql.Named("last_item_update_time", opts.withStartPageAfterItem.GetUpdateTime()),
+			sql.Named("last_item_id", opts.withStartPageAfterItem.GetPublicId()),
+		)
+	}
+
+	return r.queryRoles(ctx, query, args, limit)
+}
+
+func (r *Repository) queryRoles(ctx context.Context, query string, args []any, limit int) ([]*Role, time.Time, error) {
+	const op = "iam.(Repository).queryRoles"
+
+	var transactionTimestamp time.Time
+	var ret []*Role
+	if _, err := r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(rd db.Reader, w db.Writer) error {
+		rows, err := rd.Query(ctx, query, args)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		var inRet []*Role
+		for rows.Next() {
+			var rv Role
+			if err := rd.ScanRows(ctx, rows, &rv); err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("scan row failed"))
+			}
+			inRet = append(inRet, &rv)
+		}
```

Comment:
I think we can replace all of this with `rd.SearchWhere(ctx, &ret, whereClause, args, db.WithOrder(bla), db.WithLimit(bla))`?